initial support for '@' symbol in message template

tkrafael opened the pull request:
With this change, users will be able to create message templates containing '@' symbol.
This allows initial compatibility with https://messagetemplates.org/ and with serilog sink.
This issue [#197 ](https://github.com/serilog/serilog-extensions-logging/issues/197) can be resolved (it was closed without a proper solution).

@maryamariyan fixed an issue not related with templating, she added support for reseverd keyword parameters by prefixing them with '@' symbol (#64779)

I made those changes, tried to not break things and tried to not add performance issues. Please let me know if there's anything I have to change to allow this PR to be accepted.

Commit:
initial support for '@' symbol in message template

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -198,15 +198,15 @@ private void GenFields(LoggerMethod lm, string nestedIndentation)
                 foreach (LoggerParameter p in lm.TemplateParameters)
                 {
-                    _builder.AppendLine($"            {nestedIndentation}private readonly {p.Type} _{p.Name};");
+                    _builder.AppendLine($"            {nestedIndentation}private readonly {p.Type} _{p.CodeName};");
```
Is this going to work? wouldn't generate a compilation error when using prefix `_@`?

tarekgh replied:
@maryamariyan I appreciate if you have a few minutes to take a quick look at this one.

tkrafael replied:
Just pushed a commit to resolve this problem.

tarekgh commented:
@tkrafael looks the CI is waiting for you to accept the `license/cla` to run the tests. Could you please handle any request you are getting? or update if no action is needed from your side.

tkrafael commented:
"I have read the CLA Document and I hereby sign the CLA"

tkrafael commented:
I have read the CLA Document and I hereby sign the CLA

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79038) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79038) before we can accept your contribution.<br/><br/>:x: tkrafael [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79038)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=79038) it.</sub>

tkrafael commented:
Resulting generate file for Roslyn3.11:

``` c#
// <auto-generated/>
#nullable enable

namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
{
    partial class AtSymbolTestExtensions
    {
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, global::System.String, global::System.Exception?> __M0Callback =
            global::Microsoft.Extensions.Logging.LoggerMessage.Define<global::System.String>(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(0, nameof(M0)), "M0 {event}", new global::Microsoft.Extensions.Logging.LogDefineOptions() { SkipEnabledCheck = true }); 

        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        internal static partial void M0(global::Microsoft.Extensions.Logging.ILogger logger, global::System.String @event)
        {
            if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Information))
            {
                __M0Callback(logger, @event, null);
            }
        }
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, global::System.String, global::System.Exception?> __M1Callback =
            global::Microsoft.Extensions.Logging.LoggerMessage.Define<global::System.String>(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(1, nameof(M1)), "M1 {@myevent}", new global::Microsoft.Extensions.Logging.LogDefineOptions() { SkipEnabledCheck = true }); 

        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        internal static partial void M1(global::Microsoft.Extensions.Logging.ILogger logger, global::System.String @myevent)
        {
            if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Information))
            {
                __M1Callback(logger, @myevent, null);
            }
        }
    }
}
```
Roslyn4.0:

``` c#
// <auto-generated/>
#nullable enable

namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
{
    partial class AtSymbolTestExtensions
    {
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, global::System.String, global::System.Exception?> __M0Callback =
            global::Microsoft.Extensions.Logging.LoggerMessage.Define<global::System.String>(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(0, nameof(M0)), "M0 {event}", new global::Microsoft.Extensions.Logging.LogDefineOptions() { SkipEnabledCheck = true }); 

        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        internal static partial void M0(global::Microsoft.Extensions.Logging.ILogger logger, global::System.String @event)
        {
            if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Information))
            {
                __M0Callback(logger, @event, null);
            }
        }
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, global::System.String, global::System.Exception?> __M1Callback =
            global::Microsoft.Extensions.Logging.LoggerMessage.Define<global::System.String>(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(1, nameof(M1)), "M1 {@myevent}", new global::Microsoft.Extensions.Logging.LogDefineOptions() { SkipEnabledCheck = true }); 

        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
        internal static partial void M1(global::Microsoft.Extensions.Logging.ILogger logger, global::System.String @myevent)
        {
            if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Information))
            {
                __M1Callback(logger, @myevent, null);
            }
        }
    }
}

```

tkrafael commented:
Found that if LoggerMessageAttribute is used without loglevel, it creates a struct object, thus causing the issue described by @tarekgh 
I'll include this test and improve code to avoid this problem.

tkrafael commented:
Need more work. When generator chooses LoggerMessage.Define, it uses Format method. This will not achieve expected behaviour which is to send string template to logger infrastructure.
Let me add more fixes to this PR.

Commit:
Fix var name issue when using LoggerMessage.Define

tkrafael commented:
Please note I added comments inside code to describe where changes were made. Roslyn does not generated those comments.
Generated struct using '@' symbols:
``` c#
/// <summary> This API supports the logging infrastructure and is not intended to be used directly from your code. It is subject to change in the future. </summary>
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
[global::System.ComponentModel.EditorBrowsableAttribute(global::System.ComponentModel.EditorBrowsableState.Never)]
private readonly struct __UseAtSymbol4Struct : global::System.Collections.Generic.IReadOnlyList<global::System.Collections.Generic.KeyValuePair<string, object?>>
{
    private readonly global::System.String @myevent;
    private readonly global::System.Int32 _otherevent;

    public __UseAtSymbol4Struct(global::System.String @myevent, global::System.Int32 otherevent)
    {
        // solved here.
        this.@myevent = @myevent;
        this._otherevent = otherevent;

    }

    public override string ToString()
    {
        var otherevent = this._otherevent;
        // not creating local var as @myevent can be read directly from "this".
        return $"Force use of Struct with error, {@myevent} {otherevent}";
    }

    public static readonly global::System.Func<__UseAtSymbol4Struct, global::System.Exception?, string> Format = (state, ex) => state.ToString();

    public int Count => 3;

    public global::System.Collections.Generic.KeyValuePair<string, object?> this[int index]
    {
        get => index switch
        {
            // adds name correctly when name contains '@'
            0 => new global::System.Collections.Generic.KeyValuePair<string, object?>("@myevent", this.@myevent),
            1 => new global::System.Collections.Generic.KeyValuePair<string, object?>("otherevent", this._otherevent),
            2 => new global::System.Collections.Generic.KeyValuePair<string, object?>("{OriginalFormat}", "Force use of Struct with error, {@myevent} {otherevent}"),

            _ => throw new global::System.IndexOutOfRangeException(nameof(index)),  // return the same exception LoggerMessage.Define returns in this case
        };
    }

    public global::System.Collections.Generic.IEnumerator<global::System.Collections.Generic.KeyValuePair<string, object?>> GetEnumerator()
    {
        for (int i = 0; i < 3; i++)
        {
            yield return this[i];
        }
    }

    global::System.Collections.IEnumerator global::System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
}
```

And partial method implementation:

``` c#
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
public static partial void UseAtSymbol4(this global::Microsoft.Extensions.Logging.ILogger logger, global::Microsoft.Extensions.Logging.LogLevel level, global::System.String @myevent, global::System.Int32 otherevent, global::System.Exception ex)
{
    if (logger.IsEnabled(level))
    {
        logger.Log(
            level,
            new global::Microsoft.Extensions.Logging.EventId(3, nameof(UseAtSymbol4)),
            new __UseAtSymbol4Struct(@myevent, otherevent),
            ex,
            __UseAtSymbol4Struct.Format);
    }
}

```

When using LoggerMessage.Define:

``` c#
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, global::System.String, global::System.Exception?> __M1Callback =
    global::Microsoft.Extensions.Logging.LoggerMessage.Define<global::System.String>(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(1, nameof(M1)), "M1 {@myevent}", new global::Microsoft.Extensions.Logging.LogDefineOptions() { SkipEnabledCheck = true }); 

[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "42.42.42.42")]
internal static partial void M1(global::Microsoft.Extensions.Logging.ILogger logger, global::System.String @myevent)
{
    if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Information))
    {
        __M1Callback(logger, @myevent, null);
    }
}
```

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -603,5 +608,17 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
 
             return sb.ToString();
         }
+        private static bool ContainsAtSymbol(string value)=> value.Length > 0 && value[0] == '@';
```
>         [](http://example.com/codeflow?start=0&length=8)

nit: could you add space lines between the methods?

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -603,5 +608,17 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
             return sb.ToString();
         }
+        private static bool ContainsAtSymbol(string value)=> value.Length > 0 && value[0] == '@';
+        private static string ProtectAtSymbol(string value)
```
nit: this can be written as 

```C#
private static string ProtectAtSymbol(string value) => 
     ContainsAtSymbol(value) ? value : $"_{value}";
```

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -249,13 +249,18 @@ private void GenCases(LoggerMethod lm, string nestedIndentation)
                 foreach (LoggerParameter p in lm.TemplateParameters)
                 {
                     string name = p.Name;
+                    if(ContainsAtSymbol(p.CodeName))
```
>if [](http://example.com/codeflow?start=20&length=2)

nit: add space after `if`

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -249,13 +249,18 @@ private void GenCases(LoggerMethod lm, string nestedIndentation)
                 foreach (LoggerParameter p in lm.TemplateParameters)
                 {
                     string name = p.Name;
```
>string name = p.Name; [](http://example.com/codeflow?start=20&length=21)

can't we just make this `string name = p.CodeName` and avoid the next line check? I think when there is no `@` Name will be equal CodeName. right?

tkrafael replied:
Right, however if I change to it, this code fails:
``` c#
if (lm.TemplateMap.ContainsKey(name))
{
    // take the letter casing from the template
    name = lm.TemplateMap[name];
}
```

I'm changing this part to:

``` c#
string name = p.CodeName;

if (lm.TemplateMap.ContainsKey(p.Name))
{
    // take the letter casing from the template
    name = lm.TemplateMap[name];
}
```

This will mimic old behavior while maintaining support for '@' symbols

tkrafael replied:
Just update code, when you have time, please take a look.

tarekgh replied:
> however if I change to it, this code fails:

Why did it fail? I am seeing the code is initializing the `TemplateMap` from the template which should have CodeName. right? anyway, the current code is incorrect because you are checking `lm.TemplateMap.ContainsKey(p.Name)` and then using `name = lm.TemplateMap[name];` which can throw if `name` != `p.Name`.

tkrafael replied:
You're right.
Debugging the code (it is very hard to debug using only vscode btw), I got the following:
```c#

// CodeName = @myevent, p.Name = myevent, tm.TemplateMap[p.CodeName] = @myevent
// CodeName = otherevent, p.Name = otherevent, tm.TemplateMap[p.Name] = otherevent, tm.TemplateMap[p.CodeName] = otherevent
// CodeName = a1, p.Name = a1, tm.TemplateMap[p.Name] = A1, tm.TemplateMap[p.CodeName] = A1
```

So I think it is safe to use CodeName to find correct casing from template.

Commit:
fix code formatting

Commit:
Fix mechanism to find correct template name

tarekgh commented:
@tkrafael this change should fix the issue https://github.com/dotnet/runtime/issues/78013. right?

vpetrusevici commented:
> ```cs
> partial void LogRequestSent(RequestRequest request);
> ```

yep. you can try as `@request` also. This character will be ignored. I suppose it's related to params like `@event` becaouse you can't use just `event`.
Anyway I think it should work for both cases

vpetrusevici commented:
What's about $ caracter? It's also used like @ but to force .ToString() using instead of Json. Will it works?
It's not so critical for me, but I just think it should be supported too.

tarekgh commented:
> What's about $ caracter? It's also used like @ but to force .ToString() using instead of Json. Will it works?

We can track this to fix it later. I think fixing `@` is the priority here.

tkrafael commented:
> @tkrafael this change should fix the issue #78013. right? could you add a test case as
> 
> ```cs
> [LoggerMessage(0, LogLevel.Debug, "MMC request sent {@request}")]
> partial void LogRequestSent(RequestRequest request);
> ```
> 
> Note the parameter name not having `@` sign.
> 
> CC @vpetrusevici

Failing, checking how to solve it.

Commit:
update code with tests and more fixes

tkrafael commented:
I've added unit tests to validate parsing and code emit.
Also fixed LoggerMessage.Define when using '@' symbol, avoiding using Struct classes when possible.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -60,7 +60,8 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
                     {
                         string t = lm.TemplateList[i];
-                        if (!t.Equals(lm.TemplateParameters[i].Name, StringComparison.OrdinalIgnoreCase))
+                        var (template, parameter) = SanitizeAtSign(t, lm.TemplateParameters[i].CodeName);
```
nit: 
Can we use the span here instead of string? I mean having

```C#
        static ReadOnlySpan<char> SanitizeAtSign(ReadOnlySpan<char> template)
            => (template.Length > 0 && template[1] == '@') ? template.Slice(1) : template;
```

and do 

```
    if (!SanitizeAtSign(template).Equals(SanitizeAtSign((parameter), StringComparison.OrdinalIgnoreCase))
       ...
```

tkrafael replied:
@tarekgh just uploaded code to make use of ReadOnlySpan. Also renamed method so we can handle '$' symbols in variable too.

tarekgh reviewed (approved):
Thanks @tkrafael for the provided help here!
